pkg/agent/multicluster/policy_only_route_controller.go

```diff
@@ -0,0 +1,306 @@
+// for any Pod to forward cross-cluster traffic from Gateway to general Node
+// inside a member cluster when agent is enabled with networkPolicyOnly mode.
+type MCWithPolicyOnlyNodeRouteController struct {
+	mcClient        mcclientset.Interface
+	ovsBridgeClient ovsconfig.OVSBridgeClient
+	ofClient        openflow.Client
+	interfaceStore  interfacestore.InterfaceStore
+	nodeConfig      *config.NodeConfig
+	queue           workqueue.RateLimitingInterface
+	podInformer     cache.SharedIndexInformer
+	podLister       corelisters.PodLister
+	gwInformer      cache.SharedIndexInformer
+	gwLister        mclisters.GatewayLister
+	activeGWName    string
+	namespace       string
+	mutex           sync.Mutex
+	installedPods   map[string]string
+}
+
+type endpointChangeInfo struct {
+	isDelete  bool
+	podIP     string
+	podNodeIP string
+}
+
+func NewMCWithPolicyOnlyNodeRouteController(
+	mcClient mcclientset.Interface,
+	podInformer cache.SharedIndexInformer,
+	gwInformer cache.SharedIndexInformer,
+	client openflow.Client,
+	ovsBridgeClient ovsconfig.OVSBridgeClient,
+	interfaceStore interfacestore.InterfaceStore,
+	nodeConfig *config.NodeConfig,
+	namespace string,
+) *MCWithPolicyOnlyNodeRouteController {
+	controller := &MCWithPolicyOnlyNodeRouteController{
+		mcClient:        mcClient,
+		ovsBridgeClient: ovsBridgeClient,
+		ofClient:        client,
+		interfaceStore:  interfaceStore,
+		nodeConfig:      nodeConfig,
+		queue:           workqueue.NewNamedRateLimitingQueue(workqueue.NewItemExponentialFailureRateLimiter(minRetryDelay, maxRetryDelay), "MCWithPolicyOnlyNodeRouteController"),
+		podInformer:     podInformer,
+		podLister:       corelisters.NewPodLister(podInformer.GetIndexer()),
+		gwInformer:      gwInformer,
+		gwLister:        mclisters.NewGatewayLister(gwInformer.GetIndexer()),
+		namespace:       namespace,
+		installedPods:   make(map[string]string),
+	}
+
+	controller.gwInformer.AddEventHandlerWithResyncPeriod(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc: func(cur interface{}) {
+				controller.enqueueGateway(cur, false)
+			},
+			// Gateway UPDATE event doesn't impact Pod flows, so ignore it.
+			DeleteFunc: func(old interface{}) {
+				controller.enqueueGateway(old, true)
+			},
+		},
+		resyncPeriod,
+	)
+
+	controller.podInformer.AddEventHandlerWithResyncPeriod(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc: func(cur interface{}) {
+				controller.enqueuePod(cur, false)
+			},
+			UpdateFunc: func(old, cur interface{}) {
+				controller.enqueuePod(cur, false)
+			},
+			DeleteFunc: func(old interface{}) {
+				controller.enqueuePod(old, true)
+			},
+		},
+		resyncPeriod,
+	)
+	return controller
+}
+
+func (c *MCWithPolicyOnlyNodeRouteController) enqueueGateway(obj interface{}, isDelete bool) {
+	gw, isGW := obj.(*mcv1alpha1.Gateway)
+	if !isGW {
+		klog.ErrorS(errors.New("received unexpected object"), "enqueueGateway can't process event", "obj", obj)
+		return
+	}
+
+	if gw.Namespace != c.namespace {
```

Comment:
Is it able to use namespace as a filter when preparing gwInformer?

Reply:
Done, refined the `mcInformerFactory` in the agent.go with a `FieldSelector`.